server/src/main/java/io/druid/client/HttpServerInventoryView.java

```diff
@@ -0,0 +1,719 @@
+      Executor exec, SegmentCallback callback, Predicate<Pair<DruidServerMetadata, DataSegment>> filter
+  )
+  {
+    segmentCallbacks.put(callback, exec);
+    segmentPredicates.put(callback, filter);
+
+    finalPredicate = Predicates.or(
+        defaultFilter,
+        Predicates.or(segmentPredicates.values())
+    );
+  }
+
+  @Override
+  public void registerServerCallback(Executor exec, ServerCallback callback)
+  {
+    serverCallbacks.put(callback, exec);
+  }
+
+  @Override
+  public void registerSegmentCallback(Executor exec, SegmentCallback callback)
+  {
+    segmentCallbacks.put(callback, exec);
+  }
+
+  @Override
+  public DruidServer getInventoryValue(String containerKey)
+  {
+    synchronized (servers) {
+      return servers.get(containerKey).druidServer;
+    }
+  }
+
+  @Override
+  public Iterable<DruidServer> getInventory()
+  {
+    synchronized (servers) {
+      return Iterables.transform(
+          servers.values(), new com.google.common.base.Function<DruidServerHolder, DruidServer>()
+          {
+            @Override
+            public DruidServer apply(DruidServerHolder input)
+            {
+              return input.druidServer;
+            }
+          }
+      );
+    }
+  }
+
+  private void runSegmentCallbacks(
+      final Function<SegmentCallback, CallbackAction> fn
+  )
+  {
+    for (final Map.Entry<SegmentCallback, Executor> entry : segmentCallbacks.entrySet()) {
+      entry.getValue().execute(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              if (CallbackAction.UNREGISTER == fn.apply(entry.getKey())) {
+                segmentCallbacks.remove(entry.getKey());
+                if (segmentPredicates.remove(entry.getKey()) != null) {
+                  finalPredicate = Predicates.or(
+                      defaultFilter,
+                      Predicates.or(segmentPredicates.values())
+                  );
+                }
+              }
+            }
+          }
+      );
+    }
+  }
+
+  private void runServerCallbacks(final DruidServer server)
+  {
+    for (final Map.Entry<ServerCallback, Executor> entry : serverCallbacks.entrySet()) {
+      entry.getValue().execute(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              if (CallbackAction.UNREGISTER == entry.getKey().serverRemoved(server)) {
+                serverCallbacks.remove(entry.getKey());
+              }
+            }
+          }
+      );
+    }
+  }
+
+  //best effort wait for first segment listing fetch from all servers and then call
+  //segmentViewInitialized on all registered segment callbacks.
+  private void serverInventoryInitialized()
+  {
+    for (DruidServerHolder server : servers.values()) {
+      server.awaitInitialization();
+    }
+
+    log.info("Calling SegmentCallback.segmentViewInitialized() for all callbacks.");
+
+    runSegmentCallbacks(
```

Comment:
yes , it is for individual segments and once at inventory initialization. this naming and behavior is retained from CuratorBasedServerInventoryView ( called ServerInventoryView earlier)